catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/security/accesscontrol/AccessControlPreQueryPlugin.java

```diff
@@ -0,0 +1,246 @@
+  public void bindTag(ServiceReference metacardTypeRef) {
+    if (metacardTypeRef == null) {
+      LOGGER.trace("New reference to metacard type was null");
+      return;
+    }
+    Object tag = metacardTypeRef.getProperty(ACCESS_CONTROLLED_TAG);
+    if (tag == null) {
+      if (LOGGER.isTraceEnabled()) {
+        LOGGER.trace(
+            "Registered metacard type {} is not access-controlled", typeName(metacardTypeRef));
+      }
+      return;
+    }
+    if (!(tag instanceof String)) {
+      LOGGER.warn(
+          "The service property 'access-controlled-tag' is not valid, verify blueprint is correct");
+      return;
+    }
+    if (LOGGER.isTraceEnabled()) {
+      LOGGER.trace(
+          "Found access-controlled metacard type '{}', caching tag '{}'",
+          typeName(metacardTypeRef),
+          tag);
+    }
+    accessControlledTags.add((String) tag);
+  }
+
+  // Invoked by listener in blueprint
+  public void unbindTag(ServiceReference metacardTypeRef) {
+    if (metacardTypeRef == null) {
+      LOGGER.trace("Reference to old metacard type was null");
+      return;
+    }
+    Object tag = metacardTypeRef.getProperty(ACCESS_CONTROLLED_TAG);
+    if (tag == null) {
+      if (LOGGER.isTraceEnabled()) {
+        LOGGER.trace(
+            "Deregistered metacard type {} is not access-controlled", typeName(metacardTypeRef));
+      }
+      return;
+    }
+    if (!(tag instanceof String)) {
+      LOGGER.warn(
+          "The service property 'access-controlled-tag' is not valid, verify blueprint is correct");
+      return;
+    }
+    if (LOGGER.isTraceEnabled()) {
+      LOGGER.trace(
+          "Found access-controlled metacard type '{}', removing cached tag '{}'",
+          typeName(metacardTypeRef),
+          tag);
+    }
+    accessControlledTags.remove(tag);
+  }
+
+  // For diagnostic trace logs, see above
+  private String typeName(ServiceReference metacardTypeRef) {
+    BundleContext context =
+        FrameworkUtil.getBundle(AccessControlPreQueryPlugin.class).getBundleContext();
+    MetacardType type = (MetacardType) context.getService(metacardTypeRef);
+    return type.getName();
+  }
+
+  @Override
+  public QueryRequest process(QueryRequest input)
+      throws PluginExecutionException, StopProcessingException {
+    final String subjectIdentifier = getSubjectIdentifier();
+    final Set<String> groups = new HashSet<>(getSubjectGroups());
+    final String groupThatCanSeeEverything = configuration.getSystemUserAttributeValue();
+
+    if (groups.contains(groupThatCanSeeEverything)) {
+      LOGGER.debug(
+          "Will not modify filter; subject ({}) had at least one group [{}] that was exempt [{}]",
+          subjectIdentifier,
+          groups,
+          groupThatCanSeeEverything);
+      return input;
+    }
+
+    final TagAggregationVisitor tagVisitor = new TagAggregationVisitor();
+    final SimplifyingFilterVisitor simpleVisitor = new SimplifyingFilterVisitor();
```

Comment:
📖 By simplifying the filter before trying to trace tags against the expected result set, the entire process becomes much simpler. We no longer need to consider cases that can be logically simplified such as double negatives or redundant operators.